src/sql/src/statements/transform/expand_interval.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::ops::ControlFlow;
+
+use itertools::Itertools;
+use lazy_static::lazy_static;
+use regex::Regex;
+use sqlparser::ast::{Expr, Interval, Value};
+
+use crate::statements::transform::TransformRule;
+
+lazy_static! {
+    /// Matches either one or more digits `(\d+)` or one or more non-digits `(\D+)` characters
+    /// Negative sign before digits is matched optionally
+    static ref INTERVAL_SHORT_NAME_PATTERN: Regex = Regex::new(r"(-?\d+|\D+)").unwrap();
+
+    static ref INTERVAL_SHORT_NAME_MAPPING: HashMap<&'static str, &'static str> = HashMap::from([
+        ("y","years"),
+        ("mon","months"),
+        ("w","weeks"),
+        ("d","days"),
+        ("h","hours"),
+        ("m","minutes"),
+        ("s","seconds"),
+        ("millis","milliseconds"),
+        ("mils","milliseconds"),
+        ("ms","microseconds"),
+        ("us","microseconds"),
+        ("ns","nanoseconds"),
+    ]);
+}
+
+/// 'Interval' expression transformer
+/// - `y` for `years`
+/// - `mon` for `months`
+/// - `w` for `weeks`
+/// - `d` for `days`
+/// - `h` for `hours`
+/// - `m` for `minutes`
+/// - `s` for `seconds`
+/// - `millis` for `milliseconds`
+/// - `mils` for `milliseconds`
+/// - `ms` for `microseconds`
+/// - `us` for `microseconds`
+/// - `ns` for `nanoseconds`
+/// Required for use cases that use the shortened version of Interval declaration,
+/// f.e `select interval '1h'` or `select interval '3w'`
+pub(crate) struct ExpandIntervalTransformRule;
+
+impl TransformRule for ExpandIntervalTransformRule {
+    fn visit_expr(&self, expr: &mut Expr) -> ControlFlow<()> {
+        if let Expr::Interval(interval) = expr {
+            match *interval.value.clone() {
+                Expr::Value(Value::SingleQuotedString(value))
+                | Expr::Value(Value::DoubleQuotedString(value)) => {
+                    if let Some(data) = expand_interval_name(&value) {
+                        *expr = create_interval_with_expanded_name(
+                            interval,
+                            single_quoted_string_expr(data),
+                        );
+                    }
+                }
+                Expr::BinaryOp { left, op, right } => match *left {
```

Comment:
Great refactor! May I ask why it needs to process the `BinaryOp` left value?

Reply:
I think we don't need to expand all the string values as the left value of `BinaryOp`.

Reply:
After you've mentioned the `1y2w3d4h` case, I had to introduce some improvement 😅

> why it needs to process the BinaryOp left value?

sorry should have posted explanation comment / docs.
When I added the `SELECT INTERVAL '2h' - INTERVAL '1h'` case it was the only one failing. After a closer look I found that `Interval` AST with `BinaryOp` looks quite counter-intuitive, please find below
```sh
[src/sql/src/statements/transform/expand_interval.rs:65:9] expr.clone() = Interval(
    Interval {
        value: BinaryOp {
            left: Value(
                SingleQuotedString(
                    "2h",
                ),
            ),
            op: Minus,
            right: Interval(
                Interval {
                    value: Value(
                        SingleQuotedString(
                            "1h",
                        ),
                    ),
                    leading_field: None,
                    leading_precision: None,
                    last_field: None,
                    fractional_seconds_precision: None,
                },
            ),
        },
        leading_field: None,
        leading_precision: None,
        last_field: None,
        fractional_seconds_precision: None,
    },
)
```
strangely enough `left` is `Value::SingleQuotedString` while `right` is `Interval` (which is handled by existing logic). I was expecting both `left` & `right` to be `Interval`. That's why I had to handle the `BinaryOp` case specifically. 
The following query `SELECT INTERVAL '7 days' - INTERVAL '1d'` works with no extra(`BinaryOp`) logic.

Reply:
@etolbakov Intererting! Quite count-intuitive, thank you for sharing. I think it's necessary to handle these cases, but could we check if the right value is an interval? Try to avoid unnecessary expansion in other cases.

Reply:
Forget my last comment. The current solution looks good, thank you 👍

Reply:
@killme2008 I decided to add more tests/put comments to explain the `BinaryOp` case a bit better.